Initial-Draft.md

```diff
@@ -0,0 +1,181 @@
+```
+
+## Timeouts
+
+
+### Anti Pattern: Using arbitrary custom timeouts
+
+Lets look at the following test. The tests sets some server configurations for axp, and stubbs out a response for a merchant details endpoint. The test then waits for the response and in the then block we have an increase timeout for querying a list of results. 
+
+```js
+describe('Shop search', () => {
+  beforeEach(() => {
+  // stub out search results endpoint
+  })
+
+  describe('Affirm Anywhere merchant', () => {
+    describe('Not prequalified user', () => {
+      it('opens merchant detail page and shows Prequal VCN Flow Modal without prequal amount', () => {
+        // Sets up experiment configuratons
+        // stub out merchant details endpoint
+
+        cy.visit('/u/shop');
+        cy.findByTestId('shop-search-input').type(queryString);
+
+        cy.wait(['@searchResults']).then(() => { // cy.wait() actually uses 2 different timeouts. When waiting for a routing alias, we wait for a matching request for 5000ms, and then additionally for the server's response for 30000ms
+          cy.findByRole('listbox', { timeout: 30000 }) // query for the element with a role of `listbox` and wait ✨up to 30 seconds for it to exist in the DOM✨
+            .then(($listbox) => {  // if the listbox is found before 30 seconds is up, we go into this block
+              const firstSearchResult = $listbox.children().first();
+              return cy.wrap(firstSearchResult);
+            })
+            .click();
+        });
+      });
+    })
+  })
+});
+```
+We should keep in mind that future developers looking at this test code won't know how the `30000` was derived, increasing the timeout is only a band-aid that will lead to test flakiness, and it increases the overall runtime of integration tests in both CI and the pipeline.
+
+The above custom timeout was also added inside a `.then` block, which will only happen after the `.wait` resolves. According to cypress docs, 
+> cy.wait() actually uses 2 different timeouts. When waiting for a routing alias, we wait for a matching request for 5000ms, and then additionally for the server's response for 30000ms. That means above code will wait for ✨up to 35 seconds for the servers response✨, before the timeout will cause the test to fail. 
+
+We don't really want to give our test 30 seconds to find the listbox element. We know at this point in our code, we already have the data we need so populating it the UI should not take 30 seconds. 
+
+### Best Practice: Use Cypress assertions to wait for something to be true
+
+Lets see how we can accomplish waiting for the search results to populate the drowpdown without increasing the timeout.
+
+```js
+cy.wait(['@searchResults']).then(() => { // wait up to the default 35 seconds to get results and then continue with our test
+    cy.get('#search-results') // Now wait up to the default timeout for the dropdown to exist
+      .find('[role="option"]') // AND contain options inside of it 
```

Comment:
Shall we add `.should('exist')` here? (Just out of curiosity)

Reply:
I think that may be redundant because `.should('have.length',4);` is checking for the existence of 4 options, but maybe including it will increase readability:
```
cy.get('#search-results') 
      .find('[role="option"]')
      .should('exist')
      .and('have.length', 4) 
```      
Thoughts?

Reply:
My bad. I thought it was checking the length of its children... I don't mind either way.